### DISCUSSION TITLE
Any good way to flush out all connections

### DISCUSSION THREAD

**Discussion by psalty** [strongswan/strongswan#1567] [Q&A] [answered]
Is there any way to terminate all connections at once without restarting the charon services

Or I have to manually go through each of the connection and do swanctl -t --ike conn{n}?

**Comment by tobiasbrunner** [comment, marked as answer]
There isn't. But why not just restart it?

**Reply by psalty** [reply]
I want to keep the charon-svc in windows just like how windows IKEEXT is up and running, and that is all
By the way, charon-svc keeps crashing whenever I do stop-service "strongswan", don't have enough time to look into though.

**Reply by tobiasbrunner** [reply]
You might want to look into [VICI](https://docs.strongswan.org/docs/5.9/plugins/vici.html) and script this instead of using `swanctl`.

> By the way, charon-svc keeps crashing whenever I do stop-service "strongswan", don't have enough time to look into though.

Anything interesting logged? Did you build with `--enable-dbghelp-backtraces`?

**Reply by psalty** [reply]
I will try to build --enable-dbghelp-backtraces with it. And will let you know.
By the way, --enable-dbghelp-backtraces is supported in windows build as well right?

**Reply by tobiasbrunner** [reply]
It's only supported there.

**Reply by psalty** [reply]
--enable-dbghelp-backtraces after enable, am I supposed to see something in the charon log?

**Reply by psalty** [reply]
nothing I see it, but from the windbg I know it crashes in thread_exit in libstrongswan.
Maybe it is crashing at somewhere terminating worker thread.

**Reply by tobiasbrunner** [reply]
> --enable-dbghelp-backtraces after enable, am I supposed to see something in the charon log?

A backtrace when it crashes due to e.g. a segmentation fault or some other fatal signal.

> nothing I see it, but from the windbg I know it crashes in thread_exit in libstrongswan.

That seems very unlikely as that's only called in the unit tests, not by the daemon. Anyway, we need a full backtrace.

**Reply by psalty** [reply]
Here is the backtrack.
```
00 00000000`0c4dfb10 00007ffe`4c8bfcb4 ntdll!RtlpWaitOnCriticalSection+0xa6
01 00000000`0c4dfbf0 00007ffe`4c8bfae2 ntdll!RtlpEnterCriticalSectionContended+0x1c4
02 00000000`0c4dfc50 00000000`026c2a60 ntdll!RtlEnterCriticalSection+0x42
03 00000000`0c4dfc80 00000000`026c1c2c libstrongswan_0!rwlock_condvar_create+0xf5
04 00000000`0c4dfcb0 00000000`026c1cce libstrongswan_0!thread_exit+0x94
05 00000000`0c4dfd00 00000000`0267130d libstrongswan_0!thread_exit+0x136
06 00000000`0c4dfd30 00007ffe`4c8a9a1d libstrongswan_0+0x130d
07 00000000`0c4dfd80 00007ffe`4c8a75b6 ntdll!LdrpCallInitRoutine+0x61
08 00000000`0c4dfdf0 00007ffe`4c8e467e ntdll!LdrShutdownThread+0x166
09 00000000`0c4dfef0 00007ffe`4b70761d ntdll!RtlExitUserThread+0x3e
0a 00000000`0c4dff30 00007ffe`4c8e26a1 KERNEL32!BaseThreadInitThunk+0x1d
0b 00000000`0c4dff60 00000000`00000000 ntdll!RtlUserThreadStart+0x21
```

**Reply by tobiasbrunner** [reply]
That happens when the daemon is terminated? Does it always happen? Do you need to do anything before it happens or is starting and immediately stopping it enough?

Anyway, I currently don't see how. Again, `thread_exit()` is only called by the unit test framework. And I also don't see a path from `thread_exit()` to `rwlock_condvar_create()`. The latter is only called from the constructors of some components, never during deinitialization and definitely not from `thread_exit()`. So this backtrace doesn't make much sense. It's possible that optimization mangles it, compiling with `-O0` might give a more accurate backtrace.

**Reply by psalty** [reply]
I started charon service and stop charon service right away and it happens.

I already using -O0 option.

By the way, the backtrace is from windbg, because I could not get any backtrace from the log file or console.

Do you have instructions for me to follow?

For example, what version of dbghelp.dll do I have to use?

Besides, I don't see a clear path when calling backtrace_dump or calling backtrace_create in windows project unlike any other platform where they call it in segv_handler.

**Reply by tobiasbrunner** [reply]
> Do you have instructions for me to follow?

Sorry, no.

> For example, what version of dbghelp.dll do I have to use?

No idea whatsoever.



> Besides, I don't see a clear path when calling backtrace_dump or calling backtrace_create in windows project unlike any other platform where they call it in segv_handler.

Yeah, could definitely be that it's not possible to handle such fatal signals (or whatever the equivalent is on Windows) from within the program.

**Reply by psalty** [reply]
I started to suspect the compiler x86_64-w64-mingw32-gcc (GCC) 9.3-win32 20200320 that I am using.
After I added dumb DBG1(..) right after libcharon_deinit(), the crash was gone regardless of the optimization level.

**Reply by tobiasbrunner** [reply]
You mean in `init_and_run()`? That's weird because after `libcharon_deinit()` the daemon should be in single-threaded mode, so there shouldn't be any race conditions that a `DBG1()` could workaround. But maybe canceling and joining the worker threads doesn't work as intended. Actually, could you try if this patch makes a difference (without debug statement):

```diff
diff --git a/src/libstrongswan/threading/windows/thread.c b/src/libstrongswan/threading/windows/thread.c
index e70e4497b076..1825a789bcc7 100644
--- a/src/libstrongswan/threading/windows/thread.c
+++ b/src/libstrongswan/threading/windows/thread.c
@@ -324,8 +324,8 @@ static void end_thread(private_thread_t *this)
 	}
 	else
 	{
-		this->terminated = TRUE;
 		docleanup(this);
+		this->terminated = TRUE;
 	}
 }
```

**Reply by psalty** [reply]
It doesn't make any difference. Still crashes right after libcharon_deinit().

**Reply by tobiasbrunner** [reply]
OK, thanks for testing. I think I might have found the issue. Can you please try this:

```diff
diff --git a/src/libstrongswan/threading/windows/thread.c b/src/libstrongswan/threading/windows/thread.c
index e70e4497b076..7ac8732dbe59 100644
--- a/src/libstrongswan/threading/windows/thread.c
+++ b/src/libstrongswan/threading/windows/thread.c
@@ -637,6 +637,11 @@ static void cleanup_tls()
 	private_thread_t *this;
 	bool old;
 
+	if (!threads_lock)
+	{
+		return;
+	}
+
 	old = set_leak_detective(FALSE);
 	threads_lock->lock(threads_lock);
 
@@ -698,5 +703,6 @@ void threads_deinit()
 	destroy(this);
 
 	threads_lock->destroy(threads_lock);
+	threads_lock = NULL;
 	threads->destroy(threads);
 }
```

**Reply by psalty** [reply]
It seems like fixing the problem.

**Reply by tobiasbrunner** [reply]
OK, great. I pushed the fix to the _1567-win-thread-cleanup_ branch.
